### ISSUE TITLE
A directory traversal issue

### ISSUE THREAD

**Issue by JacksonGL** [henrytseng/hostr#8]
This package is vulnerable to directory traversal. An attacker can provide input such as ```../``` to read files outside of the specified working directory.

**Comment by henrytseng** [comment]
Can you provide a more concrete example?  I don't believe that's the case at all.

**Comment by JacksonGL** [comment]
Sure, see the concrete example below that demonstrates the vulnerability:
1. Set up a experiment directory ```A``` that contains a subdirectory ```B``` and a file ```C```. 
2. Start hostr inside directory ```B``` so that hostr serves files in the working directory (```B```).
3. Use the following node.js program to request file ```C```, which is outside the working directory ```B```:
```javascript
var http = require('http');
http.get('http://localhost:3000/../C', (res) => {
  res.on('data', (chunk) => {
    console.log(chunk.toString());
  });
});
```

You can generalize this example a little bit by replacing ```http://localhost:3000/../C``` with  some file that is more sensitive: ```http://localhost:3000/../../../etc/passwd```.  Since hostr also lists files in a directory, it basically allows an attacker to view and get any file on the hosting machine.

**Comment by henrytseng** [comment]
Yes, I know what directory traversal is.  Your example shows a basic example, but `hostr` is a simple command line utility.  It's purpose is really for development.  

The example you've showed does not use any code from `hostr`.  If you try running hostr on one of your folders you'll see that directory traversal is not possible since it simply uses the ExpressJS static module where directory traversal was fixed quite a while ago.  Perhaps you have a dependency issue?

**Comment by henrytseng** [comment]
Yes, I'm familiar with directory traversal.  Directory traversal is not possible because of the router.  Start `../bin/hostr` in the `examples/` folder and run a quick curl command:

`curl http://localhost:3000/../LICENSE`

This was a quick development tool and not intended for production obviously.  I hope you're not using it for anything production related.  If there's a bug with a specific version of node let me know (e.g. - tested on 4.2.2 and 6.9.2 LTS).  Thanks!

**Comment by JacksonGL** [comment]
Right, ```curl``` will not work probably because of the router. That is why I spend time to type the JavaScript code in my previous response. The JS program will successfully retrieve the file over the network.

If I set up hostr as my local development tool, it serves all files (not just the files in the working directory) on my machine to the local network, which is dangerous.

**Comment by henrytseng** [comment]
@JacksonGL The modules in the code were not intended to be used within a project it should be used as an external tool.  I hope that makes sense

I'm using `curl` to test so that the request is sent to the operating system.  

If your code is imports modules from `hostr` then you'd have to incorporate additional logic to prevent a tree traversal.  

If your code is external it'll serve the same response:

**Comment by JacksonGL** [comment]
Thanks for your reply. I try to tell you the vulnerability because it is published on npm and there are over 10k downloads. So I assume that there are some other people using your work as a development tool.

Not sure if I understand "using as an external tool" means precisely, I started the server using the ```hostr``` command provided by the tool. I tested the vulnerability by sending the request from both my local machine and another machine in the local network. And in both scenarios, I can successfully get any file on the hosting machine (outside the working directory).

PS: I think both ```curl``` and ```http.get``` send the request through the OS, since the ```http``` package is essentially a JS wrapper of OS system call. However, attacking from ```curl``` does not work, but attacking from the JS program actually works, I don't know why.
